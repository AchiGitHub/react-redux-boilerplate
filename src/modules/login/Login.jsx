import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter, Link } from "react-router-dom";
import { Field, reduxForm } from "redux-form";
import { CustomButton } from "./../../components/controls/CustomButton.jsx";
import { InputField } from "./../../components/controls/Fields.jsx";
// import history from "./../../_helpers/history";
// Local imports
import { loginActions, loginSelectors } from "./ducks";
import { bindActionCreators } from "redux"

// Presentational component: Login
class Login extends Component {
  constructor() {
    super();
    this.state = {
      loading: false,
      error: undefined
    };
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
  }

  static getDerivedStateFromProps(nextProps, state) {
    let newProps = {}

    if (nextProps.error !== state.error) {
      newProps.error = nextProps.error
      // newProps.loading = false

    }

    return {
      ...newProps
    }
  }

  componentWillUnmount() {
  }

  handleSubmit(values) {
    this.setState({
      // error: undefined,
      loading: true,
    });
    const { username, password } = values;
    this.props.loginActions.login({
      username,
      password
    });
  }

  render() {
    const { error, loading } = this.state;
    const { hasError, handleSubmit } = this.props;
    return (
      <div className="container card-page">
        <div className="card col-sm-12">
          <form
            className="login-form"
            onSubmit={handleSubmit(this.handleSubmit)}
          >
            <h4
              className="text-center"
              style={{ paddingBottom: 5, color: "#A9A9A9" }}
            >
              BoilerPlate Login
            </h4>
            <div className="form-group row">
              <label className="col-sm-12 col-md-4 col-form-label" required>
                Email (Username)
              </label>
              <div className="col-sm-12 col-md-8">
                <Field
                  id="username"
                  type="text"
                  className="form-control"
                  placeholder=""
                  name="username"
                  component={InputField}
                />
              </div>
            </div>
            <div className="form-group row">
              <label className="col-sm-12 col-md-4 col-form-label" required>
                Password
              </label>
              <div className="col-sm-12 col-md-8">
                <Field
                  id="password"
                  type="password"
                  className="form-control"
                  placeholder=""
                  name="password"
                  component={InputField}
                />
              </div>
            </div>
            <CustomButton
              id="login"
              disabled={loading}
              // loading={loading}
              // loading={loading && this.props.connectionDetailsLoading}
              text="Submit"
              style={{ float: "right" }}
            />

            <div className="row">
              <div className="col-sm-12">
                <span style={{ color: "#fff", fontSize: "small" }}>
                  New User?{" "}
                  {/* <Link
                    to="/customer"
                    style={{ color: "#F8921A", cursor: "pointer" }}
                  >
                    Create an Account
                  </Link> */}
                </span>
              </div>
            </div>
            <div className="row">
              <div className="col-sm-12" style={{ textAlign: "left" }}>
                <Link
                  to="/password"
                  style={{
                    color: "#fff",
                    cursor: "pointer",
                    fontSize: 10,
                    textDecoration: "underline"
                  }}
                >
                  Forgot Password?
                </Link>
              </div>
            </div>
          </form>
        </div>
      </div>
    );
  }
}
const validate = values => {
  const errors = {};
  if (!values.username) {
    errors.username = "Email is required";
  } else if (
    !/^[A-Z0-9._%+-]+@[A-Z0-9.-]+\.[A-Z]{2,4}$/i.test(values.username)
  ) {
    errors.username = "Invalid email";
  }
  if (!values.password) {
    errors.password = "Password is required";
  }
  return errors;
};
const mapStateToProps = state => {
  return {
    ...loginSelectors.getAuthenticated(state),
    ...loginSelectors.getError(state)
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    loginActions: bindActionCreators(loginActions, dispatch)
  }
}

export default withRouter(
  reduxForm({
    form: "profile",
    validate
  })(
    connect(
      mapStateToProps,
      mapDispatchToProps
    )(Login)
  )
);
