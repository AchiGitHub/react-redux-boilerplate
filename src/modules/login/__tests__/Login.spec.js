
import React from 'react';
import { mount, shallow } from 'enzyme';

import { Message } from 'semantic-ui-react'

import { Login } from '../Login';

describe('Login >> connected componenet', () => {
    let container


    it('should render the Login component', () => {
        container = shallow(<Login />)
        expect(container).toMatchSnapshot()
    });

    it('should disable submit button initially', () => {
        container = shallow(<Login />)
        expect(container.find('#submit')).toHaveLength(1)
        expect(container.find('#submit').prop('disabled')).toEqual(true)
    });

    it('should change route to /', () => {
        const mockFn = jest.fn()
        container = shallow(<Login isAuthenticated={true} history={{ push: mockFn }} />)
        expect(mockFn).toHaveBeenCalled();
    });

    it('should disable the submit button when only username is given', () => {
        const mockFn = jest.fn()
        container = shallow(<Login loginFieldChanged={mockFn} />)
        container.find('#username').simulate('change', { target: { value: 'dil' } })
        expect(mockFn).toHaveBeenCalled();
        expect(container.find('#submit').prop('disabled')).toEqual(true)
    });

    it('should login when credentials supplied', () => {
        const mockFn = jest.fn()
        const mockLogin = jest.fn()
        container = shallow(<Login loginFieldChanged={mockFn} login={mockLogin} />)
        container.find('#username').simulate('change', { target: { value: 'dil' } })
        container.find('#password').simulate('change', { target: { value: '123' } })
        expect(mockFn).toHaveBeenCalled();
        container.find('#submit').simulate('click')
        expect(mockLogin).toHaveBeenCalled();
    });

    it('should cancel username and password fields', () => {
        const mockFn = jest.fn()
        const mockLoginCancel = jest.fn()
        container = shallow(<Login loginFieldChanged={mockFn} loginCancel={mockLoginCancel} />)
        container.find('#username').simulate('change', { target: { value: 'dil' } })
        container.find('#password').simulate('change', { target: { value: '123' } })
        expect(mockFn).toHaveBeenCalled();        
        expect(container.state('password')).toEqual('123')
        container.find('#clear').simulate('click')        
        expect(container.state('password')).toEqual('')
        // expect(container.prop('username')).toEqual('')
        // expect(container.find('#password')).toEqual('')
    });

    it('should show error message', () => {
        container = shallow(<Login hasError={true} error={{ title: '', message: '' }} />)
        expect(container.find(Message).prop('negative')).toEqual(true)
    });


});


