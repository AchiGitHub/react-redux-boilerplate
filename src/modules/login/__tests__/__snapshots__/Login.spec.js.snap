// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login >> connected componenet should render the Login component 1`] = `
<Grid
  style={
    Object {
      "height": "100%",
      "marginTop": "5em",
    }
  }
  textAlign="center"
  verticalAlign="middle"
>
  <GridRow>
    <GridColumn
      style={
        Object {
          "maxWidth": 450,
        }
      }
    >
      <Card
        fluid={true}
      >
        <CardContent>
          <CardHeader>
            <Icon
              as="i"
              name="user"
            />
             Login
          </CardHeader>
          <CardDescription>
            <Form
              as="form"
              error={false}
              size="large"
            >
              <FormGroup>
                <FormField
                  control="input"
                  id="username"
                  onChange={[Function]}
                  placeholder="Username"
                  width="16"
                />
              </FormGroup>
              <FormGroup>
                <FormField
                  control="input"
                  id="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="Password"
                  value=""
                  width="16"
                />
              </FormGroup>
              <ButtonGroup
                fluid={true}
              >
                <Button
                  as="button"
                  color="blue"
                  disabled={true}
                  id="submit"
                  onClick={[Function]}
                  type="submit"
                >
                   
                  <Icon
                    as="i"
                    name="sign in"
                  />
                   Login
                </Button>
                <ButtonOr />
                <Button
                  as="button"
                  id="clear"
                  onClick={[Function]}
                >
                  Clear
                </Button>
              </ButtonGroup>
              <Divider
                hidden={true}
              />
            </Form>
          </CardDescription>
        </CardContent>
        <CardContent
          extra={true}
        >
          <Icon
            as="i"
            name="copyright"
          />
          MIT 2019
        </CardContent>
      </Card>
    </GridColumn>
  </GridRow>
</Grid>
`;
